### PR TITLE
requireVarDeclFirst: use CST

### PR DESCRIPTION
**Pull request by zxqfox**
/cc @markelog @hzoo

### CODE REVIEW

File: test/specs/rules/require-var-decl-first.js [hunk 1]
```diff
@@ -200,10 +201,6 @@ describe.skip('rules/require-var-decl-first', function() {
                     testDeclStatements(checker, '"use strict";\nvar a;', 0);
                 });
 
-                it('should not return errors for single const var declaration at top of program scope', function() {
```
Comment (markelog) [inline test/specs/rules/require-var-decl-first.js]: Why we remove those?

File: lib/rules/require-var-decl-first.js [hunk 2]
```diff
@@ -108,111 +108,24 @@
...
-            // Skip comments that occurred before the previous node
-            continue;
+function isVarDeclFirst(varDecl) {
+    var elem = varDecl.previousSibling;
+    while (elem && (elem.isComment || elem.isWhitespace ||
+        elem.type === 'VariableDeclaration' ||
+        elem.type === 'Punctuator' ||
+        elem.type === 'ExpressionStatement' && elem.expression.value === 'use strict')) {
+        // We should goes up from blocks for var declaration.
+        if (elem.type === 'Punctuator' &&
```
Comment (markelog) [inline lib/rules/require-var-decl-first.js]: Can we try a different approach with `continue` statements?

Like -
```js
if (elem.type !== 'Punctuator) {
   elem = elem.previousSibling;
   continue;
}
```

?

File: lib/rules/require-var-decl-first.js [hunk 3]
```diff
@@ -108,111 +108,24 @@
...
-            // Stop processing comments that are within multiple declarators in a single variable declaration
-            break;
-        }
-
-        if (previousNode.range[0] >= comment.range[1]) {
-            // Skip comments that occurred before the previous node
-            continue;
+function isVarDeclFirst(varDecl) {
+    var elem = varDecl.previousSibling;
+    while (elem && (elem.isComment || elem.isWhitespace ||
```
Comment (markelog) [inline lib/rules/require-var-decl-first.js]: That condition could be moved to a function right? Something like -
```js
while(somethingOrOther(elem)) {
...
```
?

**Comment by markelog** [comment]
Beautiful, with CST that thing really takes shape!

File: test/specs/rules/require-var-decl-first.js (hunk 1 above)
Comment (zxqfox) [inline test/specs/rules/require-var-decl-first.js]: We have test for const and let above. And this rule ignores them at all.

File: lib/rules/require-var-decl-first.js (hunk 2 above)
Comment (zxqfox) [inline lib/rules/require-var-decl-first.js]: I'm not sure we even need to allow `var`s inside blocks. I mean, if there is `var` — it should be placed FIRST in scope, while scope for `var` is some kind of function node or the program.

Fixed in the second commit. Please check it out!

File: lib/rules/require-var-decl-first.js (hunk 3 above)
Comment (zxqfox) [inline lib/rules/require-var-decl-first.js]: Please, look at the code. I'll move it out if need, but looks like not needed.

File: lib/rules/require-var-decl-first.js [hunk 4]
```diff
@@ -108,21 +108,22 @@
...
+        elem.type === 'BlockStatement' && (
+            elem.parentElement.type === 'FunctionExpression' ||
+            elem.parentElement.type === 'FunctionDeclaration' ||
+            elem.parentElement.type === 'ArrowFunctionExpression'
+        );
+}
+
 function isVarDeclFirst(varDecl) {
     var elem = varDecl.previousSibling;
     while (elem && (elem.isComment || elem.isWhitespace ||
```
Comment (markelog) [inline lib/rules/require-var-decl-first.js]: Much better!

How about we go further with it - 
```js
 // Probably mess up some condition
 function isRelevant /*or some better name*/ (elem) {
    var itIs = false;

    if (!elem) {
        return itIs;
    }

    if (elem.expression.value !== 'use strict') {
        return itIs;
    }

    itIs = elem.isComment || elem.isWhitespace;
    itIs = itIs || elem.type === 'VariableDeclaration';
    itIs = itIs || elem.type === 'Punctuator';
    itIs = itIs || elem.type === 'ExpressionStatement'

    return itIs;
  }

while(IsRelevant(elem)) {
  elem = elem.previousSibling;
}
```

Does it reads better?

Comment (zxqfox) [inline lib/rules/require-var-decl-first.js]: Better? ;-)

**Comment by zxqfox** [comment]
Thanks!

**Pull request closed without merging**